### PR TITLE
[HIP/AMD] #define fix & Diffusion kernel optimization

### PR DESCRIPTION
**Pull request by camierjs**
- [x] Fix `__HIP_DEVICE_COMPILE__`
- [x] Add top `backends.hpp` file for all backends
- [x] Add `MFEM_LAMBDA` vs `MFEM_DEVICE` for `hipcc` compiler
- [x] Use `MFEM_UNROLL` in `Diffusion` kernel
- [x] Fix unused variable warning

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#1588](https://github.com/mfem/mfem/pull/1588) | @camierjs | @tzanio | @v-dobrev + @acfisher + @YohannDudouit | 07/04/20 | 07/14/20 | 07/19/20 | |

### CODE REVIEW

**Comment by camierjs** [comment]
LGTM

File: general/cuda.hpp [hunk 1]
```diff
@@ -15,17 +15,15 @@
...
-#include <cuda_runtime.h>
-#include <cuda.h>
-#endif
-
 // CUDA block size used by MFEM.
 #define MFEM_CUDA_BLOCKS 256
 
 #ifdef MFEM_USE_CUDA
 #define MFEM_DEVICE __device__
+#define MFEM_LAMBDA __host__
```
Comment (YohannDudouit) [inline general/cuda.hpp:23]: I would have found more consistent to call this `MFEM_HOST`, is there a reason to choose `MFEM_LAMBDA` over `MFEM_HOST`?

File: general/cuda.hpp [hunk 2]
```diff
@@ -15,17 +15,15 @@
...
 #define MFEM_HOST_DEVICE __host__ __device__
+#define MFEM_DEVICE_SYNC MFEM_GPU_CHECK(cudaDeviceSynchronize())
+#define MFEM_STREAM_SYNC MFEM_GPU_CHECK(cudaStreamSynchronize(0))
```
Comment (YohannDudouit) [inline general/cuda.hpp:26]: Can you add some documentation to give the intuition when one should be used over the other?

File: general/hip.hpp [hunk 3]
```diff
@@ -15,16 +15,15 @@
...
-#ifdef MFEM_USE_HIP
-#include <hip/hip_runtime.h>
-#endif
-
 // HIP block size used by MFEM.
 #define MFEM_HIP_BLOCKS 256
 
 #ifdef MFEM_USE_HIP
 #define MFEM_DEVICE __device__
+#define MFEM_LAMBDA __host__ __device__
```
Comment (YohannDudouit) [inline general/hip.hpp:23]: It seems inconsistent that the hip `MFEM_LAMBDA` is defined as `__host__ __device__` whereas the cuda one is defined as `__host__`, is there a reason for that?

File: general/backends.hpp [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+#ifdef MFEM_USE_CUDA
+#include <cuda_runtime.h>
+#include <cuda.h>
+#endif
+#include "cuda.hpp"
+
+#ifdef MFEM_USE_HIP
+#include <hip/hip_runtime.h>
+#endif
+#include "hip.hpp"
```
Comment (YohannDudouit) [inline general/backends.hpp:26]: I would expect `cuda.hpp` and `hip.hpp` to not be included when Cuda and Hip are not used respectively, is it necessary to include them?

File: fem/bilininteg_mass_pa.cpp [hunk 5]
```diff
@@ -41,6 +41,8 @@ void MassIntegrator::SetupPA(const FiniteElementSpace &fes, const bool force)
       InitCeedCoeff(Q, ptr);
       return CeedPAMassAssemble(fes, *ir, *ptr);
    }
+#else
+   MFEM_CONTRACT_VAR(force);
```
Comment (YohannDudouit) [inline fem/bilininteg_mass_pa.cpp:45]: I don't understand this line, can you tell my why it is needed?

File: general/cuda.hpp (hunk 1 above)
Comment (camierjs) [reply general/cuda.hpp:23]: `nvcc` only accepts `__host__` there, where `hipcc` requires `__host__ __device__` to work.
That's why I added this variant, to be able to pick the right one from the header.

File: general/hip.hpp (hunk 3 above)
Comment (camierjs) [reply general/hip.hpp:23]: the two compilers behave differently

File: fem/bilininteg_mass_pa.cpp (hunk 5 above)
Comment (camierjs) [reply fem/bilininteg_mass_pa.cpp:45]: it removes one compiler warning when we're not using `libCeed` and compiling with `nvcc`.

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Comment by camierjs** [comment]
> Code looks good. How much performance was gained with these implementations?

There isn't that much of differences, but it allow to use less shared memory and to run the 8th order for the BP3 benchmarks.

[v100_1-7_bp3_v1.pdf](https://github.com/mfem/mfem/files/4925943/v100_1-7_diff_v1.pdf)

[v100_1-8_bp3_v2.pdf](https://github.com/mfem/mfem/files/4925944/v100_1-8_diff_v2.pdf)

File: general/cuda.hpp (hunk 2 above)
Comment (camierjs) [reply general/cuda.hpp:26]: `done`

File: general/backends.hpp (hunk 4 above)
Comment (camierjs) [reply general/backends.hpp:26]: The Cuda/Hip API (`CuMemAlloc`) is made available through these two files for now.
It's been used in the `MemoryManager`.
We could be able to hide it - that's something we can think of while refactoring.

**Comment by camierjs** [comment]
LGTM

**Comment by tzanio** [comment]
Re-merged in `next` for testing ...

**Pull request merged**